### PR TITLE
CI test automation script cp_multi-master-extra-node_ugprade.sh

### PR DESCRIPTION
**Pull request by Jaakko-Os**
Multi master part of scripts are separate
from single master on purpose.

### CODE REVIEW

File: upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+NODE_IP_LIST=()
+for node in "${ORIGINAL_NODE_LIST[@]}";do
+    NODE_IP_LIST+=$(sudo virsh net-dhcp-leases baremetal | grep "${node}"  | awk '{{print $5}}' | cut -f1 -d\/)
+done
+echo "NODE_IP_LIST ${NODE_IP_LIST[@]}"
+wait_for_ctrlplane_provisioning_complete ${ORIGINAL_NODE_LIST[@]} ${NODE_IP_LIST[@]}
+
+FROM_VERSION=$(kubectl get kcp -n metal3 -oyaml | grep "version: v1" | cut -f2 -d':' | awk '{$1=$1;print}')
+
+if [[ "${FROM_VERSION}" < "${UPGRADED_K8S_VERSION_2}" ]]; then
```
Comment (furkatgofurov7) [inline upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh:26]: `UPGRADED_K8S_VERSION_2` where is this variable defined?

Comment (Jaakko-Os) [reply upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh:26]: definition in common.sh

File: upgrade_tests/common.sh [hunk 2]
```diff
@@ -51,96 +55,176 @@ function provision_worker_node() {
...
-        sleep 1
-        if [[ "${i}" -ge 3600 ]];then
-            echo "Error: provisioning took too long to start"
-            exit 1
-            fi
-        continue
+    echo "Waiting for provisioning of controlplane node to start, number of replicas ${NUM_OF_NODE_REPLICAS}"
+    if [ "${NUM_OF_NODE_REPLICAS}" -eq 1 ];then
+        for i in {1..3600};do
+        kubectl get bmh -n metal3 | awk 'NR>1'| grep -i 'provision'
```
Comment (kashifest) [inline upgrade_tests/common.sh:61]: parameterizing the namespace would be ideal

File: upgrade_tests/common.sh [hunk 3]
```diff
@@ -51,96 +55,176 @@ function provision_worker_node() {
...
+            echo -n "."
+            break
+        fi
+        done
     else
-        echo -n "."
-        break
+        for i in {1..3600};do
+    	    provisioned_bmhs=$(kubectl get bmh -n metal3 | awk 'NR>1'| grep -i 'provision' | wc -l)
+            running_machines=$(kubectl get machines -n metal3 | awk 'NR>1'| grep 'Running' | wc -l)
```
Comment (kashifest) [inline upgrade_tests/common.sh:78]: parameterizing the namespace would be ideal

File: upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+echo "BareMetalHosts ${ORIGINAL_NODE_LIST} are in provisioning or provisioned state"
+
+NODE_IP_LIST=()
+for node in "${ORIGINAL_NODE_LIST[@]}";do
+    NODE_IP_LIST+=$(sudo virsh net-dhcp-leases baremetal | grep "${node}"  | awk '{{print $5}}' | cut -f1 -d\/)
+done
+echo "NODE_IP_LIST ${NODE_IP_LIST[@]}"
+wait_for_ctrlplane_provisioning_complete ${ORIGINAL_NODE_LIST[@]} ${NODE_IP_LIST[@]}
+
+FROM_VERSION=$(kubectl get kcp -n metal3 -oyaml | grep "version: v1" | cut -f2 -d':' | awk '{$1=$1;print}')
```
Comment (kashifest) [inline upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh:24]: parameterizing the namespace would be ideal

File: upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+echo '' > ~/.ssh/known_hosts
+
+set_number_of_node_replicas 3
+provision_controlplane_node
+
+CLUSTER_NAME=$(kubectl get clusters -n metal3 | grep Provisioned | cut -f1 -d' ')
+
+wait_for_ctrlplane_provisioning_start
+
+ORIGINAL_NODE_LIST=$(kubectl get bmh -n metal3 | grep control | grep -v ready | cut -f1 -d' ')
```
Comment (kashifest) [inline upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh:14]: parameterizing the namespace would be ideal

File: upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+
+source ../common.sh
+
+echo '' > ~/.ssh/known_hosts
+
+set_number_of_node_replicas 3
+provision_controlplane_node
+
+CLUSTER_NAME=$(kubectl get clusters -n metal3 | grep Provisioned | cut -f1 -d' ')
```
Comment (kashifest) [inline upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh:10]: parameterizing the namespace would be ideal

File: upgrade_tests/common.sh (hunk 2 above)
Comment (Jaakko-Os) [reply upgrade_tests/common.sh:61]: True. Improvements shall be implemented next after this PR is merged.

File: upgrade_tests/common.sh (hunk 3 above)
Comment (Jaakko-Os) [reply upgrade_tests/common.sh:78]: True. Improvements shall be implemented next after this PR is merged.

File: upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh (hunk 4 above)
Comment (Jaakko-Os) [reply upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh:24]: True. Improvements shall be implemented next after this PR is merged.

File: upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh (hunk 5 above)
Comment (Jaakko-Os) [reply upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh:14]: True. Improvements shall be implemented next after this PR is merged.

File: upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh (hunk 6 above)
Comment (Jaakko-Os) [reply upgrade_tests/controlplane_upgrade/cp_multi-master-extra-node_ugprade.sh:10]: True. Improvements shall be implemented next after this PR is merged.

**Comment by smoshiur1237** [comment]
Overall the PR Looks good to me as parameterize of namespace can be done later after the merge.

**Comment by Jaakko-Os** [comment]
rebased

**Pull request merged**